mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp

```diff
@@ -624,92 +624,67 @@ struct ThreadwiseWriteAllRewritePattern
+
+  assert(!sourceNeeds64BitIdx && "Registers shouldn't need 64-bit indexing");
+  assert(!dstNeeds64BitIdx && "Registers shouldn't need 64-bit indexing");
   ArrayRef<int64_t> rawLoadBufferShape =
       rawLoadBuffer.getType().cast<ShapedType>().getShape();
-
-  Value rawStoreBuffer;
-  ArrayAttr storeBufferView;
-  std::tie(rawStoreBuffer, storeBufferView, needs64BitIdx) =
-      untransform(b, destView);
-  assert(!needs64BitIdx && "Registers shouldn't need 64-bit indexing");
   ArrayRef<int64_t> rawStoreBufferShape =
       rawStoreBuffer.getType().cast<ShapedType>().getShape();
-
+  if (rawLoadBufferShape.size() != 1)
+    return op.emitOpError("Raw load buffers have to be flat.");
+  if (rawStoreBufferShape.size() != 1)
+    return op.emitOpError("Raw store buffers have to be flat.");
   Value zero = b.createOrFold<arith::ConstantIndexOp>(loc, 0);
-  SmallVector<Value, 2> start(2, zero);
-  SmallVector<int64_t, 2> strides(2, 1);
-  int64_t vecLen = 1;
-  // The store buffer is a flattened < kouter x dPerThread x kpack >.
-  if (kpackPerThread == 1) {
-    // if kpack == 1, then we can do vectorized loads across d dimension from/to
-    // load/store buffer
-    vecLen = getMaxVectorizationForDatatype(loadBufferView, /*dim=*/1,
-                                            copyDPerThread, rawLoadBufferShape,
-                                            elemType);
-    vecLen = math_util::gcd(copyDPerThread, vecLen);
-    strides[1] = vecLen;
-  } else {
-    // if kpack > 1, then we are limited by vectorization in k dimension and it
-    // could be at most kpack.
-    vecLen = getMaxVectorizationForDatatype(loadBufferView, /*dim=*/0,
-                                            copyKPerThread, rawLoadBufferShape,
-                                            elemType);
-    vecLen = math_util::gcd(vecLen, kpackPerThread);
-    strides[0] = vecLen;
-  }
-  loadBufferView = collapseContiguousMerges(loadBufferView, rawLoadBufferShape);
-  storeBufferView =
-      collapseContiguousMerges(storeBufferView, rawStoreBufferShape);
+
+  auto storeBufferViewInverted = invertTransforms(b, loc, storeBufferView);
+  auto srcToDstView =
+      prependUpperViews(b, storeBufferViewInverted, loadBufferView);
+
+  Type elemType = sourceView.getType().cast<MemRefType>().getElementType();
+  int64_t maxVlen = 128 / elemType.getIntOrFloatBitWidth();
+  int64_t vecLen = getMaxVectorizationForDatatype(
+      srcToDstView, /*dim=*/0, maxVlen, rawStoreBufferShape, elemType);
+
+  srcToDstView = collapseContiguousMerges(srcToDstView, rawStoreBufferShape);
 
   // Run the packing loop
-  auto packLoop = b.create<TransformingForOp>(
+  SmallVector<Value, 2> start{zero};
+  SmallVector<int64_t, 2> strides{vecLen};
+  auto copyLoop = b.create<TransformingForOp>(
       loc, ArrayRef<ValueRange>{start, start},
-      ArrayRef<Attribute>{loadBufferView, storeBufferView},
-      /*bounds=*/loadShape,
+      ArrayRef<Attribute>{srcToDstView, b.getArrayAttr({})},
```

Comment:
I think we'll want to still want to use two domains in the uninvertable case? Or we'll want to ban the uninvertable case?

Reply:
What I can do is to not vectorize the uinvertible case and copy from/to the original views one element at the time

Reply:
Makes sense, sounds like a plan

I think you'll just need to fix `invertTransforms` to report the failure cleanly

Reply:
I think it just returns null ... does it not ?

Reply:
No, I don't think it does, or at least it wasn't clear that it did that when I looked yesterday

Reply:
Yeah, so it was still returning a non empty array attr. I made it so that if the inversion fails, it now return an empty ArrayAttr(`{}`).

Reply:
... that's not actually an empty array attr, which is `b.getArrayAttr({})` or `ArrayAttr::get(ctx, {})` or the like. It's actually `ArrayAttr{}` aka `ArrayAttr(nullptr)`.

Which is good as a failure state.